src/backend/gl/src/device.rs

```diff
@@ -464,73 +463,81 @@ impl d::Device<B> for Device {
 
-            let raw = gl.create_buffer().unwrap();
-            //TODO: use *Named calls to avoid binding
-            gl.bind_buffer(target, Some(raw));
-            if self.share.private_caps.buffer_storage {
-                let flags = if is_device_local_memory {
-                    0
-                } else {
-                    glow::MAP_READ_BIT
-                        | glow::MAP_WRITE_BIT
-                        | glow::MAP_PERSISTENT_BIT
-                        | glow::DYNAMIC_STORAGE_BIT
-                };
-                gl.buffer_storage(target, size as i32, None, flags);
-            } else {
-                let usage = if is_device_local_memory {
-                    glow::STATIC_DRAW
+                let raw = gl.create_buffer().unwrap();
+                //TODO: use *Named calls to avoid binding
+                gl.bind_buffer(target, Some(raw));
+                if self.share.private_caps.buffer_storage {
+                    let flags = if is_device_local_memory {
+                        0
+                    } else {
+                        assert!(is_cpu_visible_memory);
+                        let read_flag = if is_readable_memory {
+                            glow::MAP_READ_BIT
+                        } else {
+                            0
+                        };
+                        read_flag
+                            | glow::MAP_WRITE_BIT
+                            | glow::MAP_PERSISTENT_BIT
+                            | glow::DYNAMIC_STORAGE_BIT
+                    };
+                    gl.buffer_storage(target, size as i32, None, flags);
                 } else {
-                    glow::DYNAMIC_DRAW
-                };
-                gl.buffer_data_size(target, size as i32, usage);
-            }
-            gl.bind_buffer(target, None);
+                    let usage = if is_device_local_memory {
+                        glow::STATIC_DRAW
+                    } else {
+                        glow::DYNAMIC_DRAW
+                    };
+                    gl.buffer_data_size(target, size as i32, usage);
+                }
+                gl.bind_buffer(target, None);
 
-            if let Err(err) = self.share.check() {
-                panic!("Error allocating memory buffer {:?}", err);
-            }
+                if let Err(err) = self.share.check() {
+                    panic!("Error allocating memory buffer {:?}", err);
+                }
 
-            let mut map_flags = glow::MAP_READ_BIT | glow::MAP_WRITE_BIT | glow::MAP_PERSISTENT_BIT;
-            if !is_coherent_memory {
-                map_flags |= glow::MAP_FLUSH_EXPLICIT_BIT;
+                let mut map_flags = glow::MAP_WRITE_BIT
+                    | glow::MAP_PERSISTENT_BIT
+                    | glow::MAP_FLUSH_EXPLICIT_BIT;
+                if is_readable_memory {
+                    map_flags |= glow::MAP_READ_BIT;
+                }
+                if !is_coherent_memory {
```

Comment:
the condition appears wrong

Reply:
Oh!  It definitely is wrong, it used to control the explicit flush bit before I figured out that it should control the coherent flag instead, and I forgot to reverse it.

Reply:
And it was hiding another bug, which is now fixed 😥